### PR TITLE
Fix EZP-26175: Exception instead of 404 when trying to access a file …

### PR DESCRIPTION
**Pull request by glye**
…stored with a non default vardir

> Fixes https://jira.ez.no/browse/EZP-26175
> Status: Ready for review

If you set the var directory to `var/storage` you get an unhandled exception. Using the default VarDir you get a 404, which makes more sense.

Got a working solution now: Throw a Symfony `NotFoundHttpException` from the `StreamFileListener`, this results in a 404 for the end user. It also means that the `MissingBinaryFile` check in `BinaryStreamResponse` is no longer needed, making it cleaner.

- [x] ~~Test the new exception classes~~ Irrelevant after last revert
- [x] ~~Document `throws`~~ Irrelevant after last revert
- [x] ~~Translate the 404 error message~~ Irrelevant after last revert
- [x] ~~Update script for existing content~~ See https://jira.ez.no/browse/EZP-26416

### CODE REVIEW

**Comment by glye** [comment]
Wow, it passes. The legacy backend works fine in my manual test, serving images and files as it should. On the symfony frontend, `Core/IO/UrlDecorator/Prefix` throws `InvalidBinaryFileIdException`: Argument 'BinaryFile::id' is invalid: 'var/ezdemo_site/storage/images/getting-started/selected-features/getting-started-with-ez-publish-platform/322-1-eng-GB/Getting-Started-with-eZ-Publish-Platform.jpg' is wrong value in class 'BinaryFile'.

@bdunogier have a looksie please, and see question in PR description.

**Comment by bdunogier** [comment]
> Question: Why is the ezpublish.core.io.stream_file_listener set to use the ezpublish.fieldtype.ezimage.io_service? Shouldn't there be some mechanism for choosing between different IO services according to the URL requested, or smth like that?

This makes sense. Not sure how we ended up with this setup, but it is clear that it can't work for both cases in the current state.

**Comment by glye** [comment]
@bdunogier Interestingly, using `@ezpublish.core.io.service` is the same as using `@ezpublish.fieldtype.ezbinaryfile.io_service` - meaning, it works. Files and images load correctly.

**Comment by bdunogier** [comment]
Hmmm, the binaryfile one loads images ?

> The exception makes sense, it doesn't find the "images" prefix. Either binary file loading will work, or images, depending on which one we set as IO Service in the StreamFileListener.

This seemed to say otherwise.

**Comment by glye** [comment]
> Either binary file loading will work, or images, depending on which one we set as IO Service in the StreamFileListener.

@bdunogier That's what I thought before actually testing it. As it turns out, both the generic and the file IO services work fine for images and files, while the image one only works for images.

**Comment by joaoinacio** [comment]
FYI: While testing this change in the context of related EZP-26244 (serving image aliases from cluster) on 5.x, I am greeted with a `Argument '$id' is invalid: Prefix original not found in ...` exception.

**Comment by glye** [comment]
@joaoinacio That's essentially the same error as reported by the creator of this issue, except for images rather than files. And that's expected given the change as it is now. I have tested some more, here are the extended results:

**Using @ezpublish.fieldtype.ezimage.io_service (i.e. before PR)**
Works:
/var/storage/storage/images/ezp-26175-real-image/877-1-eng-GB/EZP-26175-real-image_imagefull.png
/var/ezdemo_site/storage/images/ez-mountains/mount-fuji/569-1-eng-GB/Mount-Fuji.jpg
/Admin/EZP-26175-real-image

Fails:
/var/storage/storage/original/application/98fdbc938a3664c55486dc94391e9d42.pdf - `Argument '$id' is invalid: Prefix images not found`
/var/storage/storage/original/image/6ce6332e584f189c9852dcae6c58af7b.png - `Argument '$id' is invalid: Prefix images not found`
/eZ-Mountains - `InvalidBinaryFileIdException: Argument 'BinaryFile::id' is invalid`

**Using @ezpublish.core.io.service or @ezpublish.fieldtype.ezbinaryfile.io_service**
Works:
/var/storage/storage/original/application/98fdbc938a3664c55486dc94391e9d42.pdf
/var/storage/storage/original/image/6ce6332e584f189c9852dcae6c58af7b.png
/var/storage/storage/images/ezp-26175-real-image/877-1-eng-GB/EZP-26175-real-image_imagefull.png
/var/ezdemo_site/storage/images/ez-mountains/mount-fuji/569-1-eng-GB/Mount-Fuji.jpg
/Admin/EZP-26175-real-image

Fails:
/eZ-Mountains - `InvalidBinaryFileIdException: Argument 'BinaryFile::id' is invalid`

Note: /eZ-Mountains, i.e. the symfony frontend, fails in both cases.

**Comment by joaoinacio** [comment]
@glye:

Not sure I fully understand, but what I meant is that, when accessing an image such as 
`/var/storage/storage/images/ezp-26175-real-image/877-1-eng-GB/EZP-26175-real-image_imagefull.png`,
it will fail with a `Prefix original not found` - is that normal?

**Comment by andrerom** [comment]
If I get this correctly, the cluster code needs to be able to pick right IO based on patch like other parts does already. Is that it @bdunogier?

**Comment by andrerom** [comment]
Patch -> path

**Comment by glye** [comment]
@joaoinacio Normal? It's what currently happens - when you use the peculiar VarDir setting described in the issue, of course.

@andrerom So it appears. What other parts are you thinking about?

**Comment by joaoinacio** [comment]
@glye: sorry, should have been more detailed on my first comment:
Using default VarDir, Clustered 5.4, accessing an image(alias).

I believe it would be resolved by using the right IO as mentioned.

**Comment by bdunogier** [comment]
FieldTypes IO services (binaryfile, image), have a built-in prefix. One way to handle this situation is to indeed match on a given URI prefix, and use the appropriate service.

An alternative I sugges earlier in this thread, is to try the "generic" IO service, that doesn't have a fieldtype specific storage prefix. Maybe it is the cleanest way to cover this.

**Comment by glye** [comment]
In my [test](https://github.com/ezsystems/ezpublish-kernel/pull/1773#issuecomment-250395804) the binaryfile IO service behaves the same as the generic one. They work in the backend, but fail in the frontend.

Note: This bug is not cluster specific.

**Comment by glye** [comment]
Update: Using the generic **@ezpublish.core.io.service** does indeed work in all cases. We just need an update script for existing content, to be run whenever the VarDir is changed.

**Comment by glye** [comment]
Input from @bdunogier: Maybe the image IO service is after all the correct one. The binary file one would allow you to bypass permission checks for files, one should use the dedicated route for that.

Either way, an upgrade script seems still to be needed.

**Comment by glye** [comment]
Update: Going back to the original `@ezpublish.fieldtype.ezimage.io_service` and instead looking into the original complaint: That a modified VarDir results in an unhandled exception, rather than a 404. The change ensures that a `MissingBinaryFile` is created in this case, and prevents the `BinaryStreamResponse` from checking the mtime of the nonexisting file.

Referring to the **Incomplete var directory support** feature by @lolautruche:
https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/specifications/io/incomplete_vardir_handling.md
https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/bc/changes-5.4.md (see TolerantIOService)
and https://github.com/ezsystems/ezpublish-kernel/blob/196bd047bd3d9e851e7a125d331c2e459f8792ce/eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php#L58

This implies that a `MissingBinaryFile` should result in a 404, but instead I get an uncaught `File not found at path: images` within FlySystem. It appears that the feature is not complete, I mean I would not expect any dealings with FlySystem in the case of a `MissingBinaryFile`. Quite possibly I'm doing it wrong :)

CC @bdunogier

**Comment by glye** [comment]
Thanks for the approval, but this is still work in progress :) Description updated to make it clearer.

This would benefit from a look by @lolautruche (and @bdunogier)

**Comment by glye** [comment]
> @andrerom changed the base branch to 6.5 from 6.3 16 days ago

Um... why, when this is legacy specific?

**Comment by andrerom** [comment]
> > @andrerom changed the base branch to 6.5 from 6.3 16 days ago
> Um... why, when this is legacy specific?

Does not compute. 1. this is kernel repo. 2. until v1.5.2 is out, oldest supported 6.x branch is 6.5.

**Comment by glye** [comment]
> Does not compute. 1. this is kernel repo. 2. until v1.5.2 is out, oldest supported 6.x branch is 6.5.

@andrerom 1. Yes, the bug seems to be in kernel, though it manifests only in legacy installs. 2. Ack, my bad.

I think I'll close this, and move to other repo.

**Comment by andrerom** [comment]
Maybe hold for @bdunogier to have input next week, there are some issues in platform as well in regards to missing files.

**Comment by glye** [comment]
Update: Got a working solution now, see description.

**Comment by glye** [comment]
I'm not sure the fixup commit https://github.com/ezsystems/ezpublish-kernel/pull/1773/commits/624e06b504c4246716525ca136a6cba9c6bb04f1 is a good idea. It complicates things. There are existing cases like https://github.com/ezsystems/ezpublish-kernel/blob/624e06b504c4246716525ca136a6cba9c6bb04f1/eZ/Publish/Core/MVC/Symfony/Controller/Content/ViewController.php#L162 where the `NotFoundHttpException` is used directly, without translation. Maybe better to keep it like that.

**Comment by glye** [comment]
Update: Reverted the fixup, to keep things simple.

File: eZ/Publish/Core/IO/TolerantIOService.php [hunk 1]
```diff
@@ -105,6 +113,20 @@ public function loadBinaryFileByUri($binaryFileUri)
     }
 
     /**
+     * @param $binaryFileUri
+     *
+     * @return MissingBinaryFile
+     */
+    private function createMissingBinaryFileByUri($binaryFileUri)
```
Comment (bdunogier) [inline eZ/Publish/Core/IO/TolerantIOService.php]: I have a feeling that we don't need that. Creating a value object is a really simple operation, and the arguments speak for themselves.

Comment (glye) [inline eZ/Publish/Core/IO/TolerantIOService.php]: @bdunogier Ok, sure. Then I take it we don't need `createMissingBinaryFile()` either.

Comment (bdunogier) [inline eZ/Publish/Core/IO/TolerantIOService.php]: Well, `createMissingBinaryFile` does two things (create & load the URI). It makes a _little_ more sense to have a private. I don't mind if you move the logic from that private to the method itself.

**Comment by glye** [comment]
Update: Removed private method as requested, plus its sibling (separate commit).

**Comment by bdunogier** [comment]
Approved, please go ahead for the 5.4 backport, and send it to QA, please @glye.

**Pull request merged**